Fixes #6873 - MEO/PEO handling in `chordal.py`

Luis-Varona opened the pull request:
As per #6873, we extract the MCS-M triangulation logic into `minimal_elimination_order` and ensure that MCS-M is run even if the input graph is already chordal. We keep `complete_to_chordal_graph` for backwards compatibility, simply returning early if the input graph is chordal (but with the caveat that the `alpha` output will be all zeros, which is what motivated this issue/PR in the first place) and calling `minimal_elimination_order` otherwise. We also create a new function `perfect_elimination_order` which assumes that the input graph is already chordal, raising an error if it is not.

For some context, @dschult's original comments were:

> So, it seems that we have many functions that could all use this underlying code:
> 
> * [Edit: `triangulate` might be a better name] make_perfect_elimination_graph  : return the graph created in current code
> * `minimal_elimination_order` : return the order given by alpha from the current code
> * `perfect_elimination_order` : if not is_chordal raise, else return MEO given by alpha
> * `is_perfect_elimination_graph` : `is_chordal`   (no new function needed... perhaps add a sentence to docstring.
> 
> Maybe the change to `complete_to_chordal_graph` would need to be deprecated..?? The old behavior was to return alpha with all zeros. Can we just change that? It is not a useful return value. Do we need to deprecate?
> 
> If we do, let's copy the existing code to a private function, remove the is_chordal check and make a new function with the old name and doc_string that does the check and returns alpha with zeros. Then the other functions can call the private function. I suppose we could make a new public function that returns both the graph and alpha. What would a better name for that function be? The current function certainly does much more than making a graph chordal. :}

**fixes #6873**

Luis-Varona commented:
@amcandio @dschult, just want to know what other changes (if any) I should make besides the (not yet added) docstrings. Alternatively, if @gulshan-123's PR is already satisfactory, I totally don't mind if you just merge that in 🙂 No rush, of course; take your time.

dschult reviewed (commented):
Let's remove the function `is_perfect_elimiation_graph`.  We'll just have one name: `is_chordal` and we will need to make the doc_strings robust enough so the user knows that "chordal" means the graph has a "perfect elimination ordering".  I'd also like to change the name of `_triangulate_to_perfect_elimination_graph`. Perhaps `triangulate`? But maybe `triangulation` or `fully_triangulate` or ...   :}

Also, can you make sure that each of these function only does the work once?  That is, the function shouldn't call `is_chordal(G)` and then `_triangulate_to_perfect...`. The triangulation is already done in `is_chordal`.  Let's avoid that step.  Calculate it once and use that info for whatever we need.

Thanks!

Luis-Varona commented:
@dschult I've made some of the changes in this commit:

```
    Remove is_PEG alias for is_chordal; update triangulation
    
    Removed the is_perfect_elimination_graph alias for is_chordal, opting to add a sentence to the is_chordal docstring instead. (Reference needed for that, maybe?) Further, renamed the _triangulate_to_perfect_elimination_graph helper function to _triangulate_fully, and streamlined some of the logic. Finally, updated the module's __all__ list with the new functions.
    
    Still need to add docstrings, and maybe consider redundancies between is_chordal and _triangulate_fully (not sure 
about the latter?).
```

What exactly do you mean by this?

> Also, can you make sure that each of these function only does the work once? That is, the function shouldn't call `is_chordal(G)` and then `_triangulate_to_perfect...`. The triangulation is already done in `is_chordal`. Let's avoid that step. Calculate it once and use that info for whatever we need.

AFAIK, `is_chordal` isn't related to (the recently renamed) `_triangulate_fully` helper, but I could be wrong. Thoughts? 🙂

dschult commented on networkx/algorithms/chordal.py:
```diff
@@ -411,33 +413,68 @@ def complete_to_chordal_graph(G):
-    unnumbered_nodes = list(H.nodes())
-    for i in range(len(H.nodes()), 0, -1):
+
+    nodes = H.nodes()
+    weights = dict.fromkeys(nodes, 0)
+    unnumbered_nodes = list(nodes)
+    n = len(unnumbered_nodes)
```
No need to create multiple objects to hold the nodes.
```suggestion
    weights = dict.fromkeys(H, 0)
    unnumbered_nodes = list(H)
    n = len(unnumbered_nodes)
```

dschult reviewed (commented):
Yes -- I think you are right that the algorithm in `is_chordal` is not related to the algorithm in `_triangulate_fully`.  Sorry to the probably confusing comment/question. I was thinking/hoping that the current algorithm was a superset of the `is_chordal` algorithm. But it doesn't look like it.

Let's add a reference to the Wikipedia page on Chordal_Graphs. You can cite it in the comment you added.

As for comments/questions: 
Why did you change the name of `weight` to `weights`? I'm thinking that a dict's name typically describes the values -- though sometimes we name it based on the keys. When it is named for the keys, it is naming a collection of them, so is plural. When the name is describing the values it is describing what is returned upon lookup in the dict. That is usually singular (one value for each key) -- but could be plural if there are multiple objects in the dict's value. But maybe you had a different (and valid) reason for changing weight to weights.
And why change the for loop counting variable from `i` to `k`?

Some other comments below.

Luis-Varona commented:
@dschult, perhaps you are right regarding `weight`/`weights`—I'll change it back. As for changing `i` to `k` in the loop, I refactored the logic so that we're iterating forward in `i` rather than backward and added an `n - i`  assignment in the loop body. I'm just so used to seeing `n - k` in combinatorics stuff than i changed it to `k` instead, but if you have any misgivings, I can change it back too.

Regarding your comments on the reference in the `is_chordal` docstring, I'll get to that tomorrow. Will review your code suggestion tomorrow as well. Thanks for the review! 🙂 Hopefully after I make those changes and add the docstrings, it'll be ready to merge.

Luis-Varona commented:
@dschult, I've made the requested changes. Additionally, I noticed some inconsistencies in the way `References` sections were formatted as I went to add one to `is_chordal`, so my obsessive Political Science major self decided to reformat the references throughout the file in Chicago-style format, if that's okay. 🙂 Will get to work on the docstrings for the newly added functions&mdash;after that, do we expect it'll be ready to merge?

Luis-Varona commented:
@dschult @amcandio I am, subject to your approval, done with the PR and ready to merge. Let me know if you have any further comments and/or would like me to change anything else. 🙂

Luis-Varona commented on networkx/algorithms/chordal.py:
```diff
@@ -395,49 +408,126 @@ def complete_to_chordal_graph(G):
+        z = max(unnumbered_nodes, key=lambda node: weights[node])
         unnumbered_nodes.remove(z)
-        alpha[z] = i
+        alpha[z] = n - k
```
Nope, this reproduces the original behavior:
```
>>> n = 6
>>> for i in range(n, 0, -1):
...     print(i)
... 
6
5
4
3
2
1
>>> for k in range(n):
...     print(n - k)
... 
6
5
4
3
2
1
```

Luis-Varona commented on networkx/algorithms/chordal.py:
```diff
@@ -31,10 +33,12 @@ class NetworkXTreewidthBoundExceeded(nx.NetworkXException):
 @not_implemented_for("multigraph")
 @nx._dispatchable
 def is_chordal(G):
+    # TODO: Add source for `chordal ⟺ has PEO`?
```
Yup! Sorry, hehe

Luis-Varona commented on networkx/algorithms/chordal.py:
```diff
@@ -395,49 +408,126 @@ def complete_to_chordal_graph(G):
+    else:
+        H, alpha = _triangulate_fully(G)
+
+    return H, alpha
```
If `G` is chordal, simply make a copy and fill `alpha` with zeroes. If `G` is not chordal, call the `_triangulate_fully` helper containing the MCS-M algorithm. This does the same thing as our original `complete_to_chordal_graph` function, except it extracts the MCS-M logic into the `_triangulate_fully` helper so that `minimal_elimination_order` and `perfect_elimination_order` can call it as well.

Luis-Varona commented:
@amcandio @dschult I've made the last of the required changes (I hope, at least). Let me know if anything further is needed before merging, and thanks again for the feedback! 🙂

dschult commented on networkx/algorithms/chordal.py:
```diff
@@ -137,13 +143,13 @@ def find_induced_nodes(G, s, t, treewidth_bound=sys.maxsize):
-    .. [1] Learning Bounded Treewidth Bayesian Networks.
-       Gal Elidan, Stephen Gould; JMLR, 9(Dec):2699--2731, 2008.
-       http://jmlr.csail.mit.edu/papers/volume9/elidan08a/elidan08a.pdf
+    .. [1] Elidan, Gal and Stephen Gould. "Learning Bounded Treewidth Bayesian
```
Why does this have one author written as `<last name>, <first name>` and the other written as `<first name> <last name>`

Also, I much prefer, e.g., the title to avoid a line break so I can cut/paste. Having the title on one line and then the authors on the second line was IMO better.  And I'm sure everyone has a different "better". But can you give a reason that this is better than the original whitespace and order?  
And the final page number seem to have lost a leading  digit. And we've lost the information "Dec" 

The new url seems to work, but the old one worked too.  Is there a reason for changing any of this reference?

Luis-Varona replied:
> Why does this have one author written as `<last name>, <first name>` and the other written as `<first name> <last name>`

@dschult It's just standard Chicago style (CMOS), which `networkx` officially recommends (see https://networkx.org/documentation/latest/developer/contribute.html#adding-references). In the case of 3 authors, for instance, CMOS requires `LastName1, FirstName1, FirstName2 LastName2, and FirstName3 LastName3`.

> Also, I much prefer, e.g., the title to avoid a line break so I can cut/paste. Having the title on one line and then the authors on the second line was IMO better. And I'm sure everyone has a different "better". But can you give a reason that this is better than the original whitespace and order? And the final page number seem to have lost a leading digit. And we've lost the information "Dec"

In my newly added References I'll try keeping the title on one line and the authors on another, but I should probably still put the authors first in accordance with CMOS (which, again, `networkx` officially espouses). I think it's just best that I revert the changes to the old References and just fix them in a new PR. Also, CMOS does not include publication months when citing journal articles.

> The new url seems to work, but the old one worked too. Is there a reason for changing any of this reference?

Once again, CMOS asks for DOIs (which are permanent identifiers) rather than any potentially changeable URLs linking to the article. (There is just a slew of mistakes in the formatting of the old References, but again, I should probably leave that for another PR.)

dschult commented on networkx/algorithms/chordal.py:
```diff
@@ -395,35 +407,110 @@ def complete_to_chordal_graph(G):
+    return _triangulate_fully(G)
+
+
+def _triangulate_fully(G):
+    # TODO: Add docstring
```
Is there any reason not to have `_triangulate_fully` be named `minimal_elimination_order` with its doc_string? That would perhaps make this slightly easier to read. But maybe I'm missing something.

amcandio replied:
+1 to me it seems like we could just have 1 method instead of 3

Luis-Varona replied:
@dschult You're quite right; I'll just move the `_triangulate_fully` MCS-M logic into `minimal_elimination_order`. However, @amcandio, we still need all 3 of `complete_to_chordal_graph` (for backwards compatibility), `minimal_elimination_order` (for when we don't want `alpha` to be all zeros if the input is already chordal), and `perfect_elimination_order` (if we know our input is already chordal and want to raise an error if it is not).

dschult commented on networkx/algorithms/chordal.py:
```diff
@@ -283,9 +289,15 @@ def chordal_graph_treewidth(G):
 
+    Notes
+    -----
+    See [1]_ for a definition of and further information on treewidth.
```
How long is the definition of tree width? Can we summarize it here and *then* say "for further info, see..."?

dschult commented on networkx/algorithms/chordal.py:
```diff
@@ -395,35 +407,110 @@ def complete_to_chordal_graph(G):
+def perfect_elimination_order(G):
+    """Return a copy of G triangulated to a perfect elimination ordering.
+
+    A graph has a perfect elimination ordering if and only if it is chordal
```
This doc-string should include a brief definition of what a perfect elimination order is.  Referring to `is_chordal` is fine. But its better if the reader has all the relevant info in each function rather than having to chase around to find it.

Luis-Varona replied:
Sounds good! Will get on that 🙂

dschult commented on networkx/algorithms/chordal.py:
```diff
@@ -395,35 +407,110 @@ def complete_to_chordal_graph(G):
-        alpha[z] = i
+        alpha[z] = n - k
         update_nodes = []
+
```
Why these added blank lines?

dschult reviewed (commented):
Let's remove the chicago style reference changes. That seems to just add noise that will inevitably (at least with me around) lead to less than useful discussion. (I like to have it informative and convenient rather than consistent.)  It certainly distracts me when reviewing. I could imagine a separate PR that just improves references though.

Similarly -- why change all the "self loop" to "self-loop"? 
And adding periods to the end of the first line of a doc_string just makes the string longer so it may not wrap as easily on some devices. Is there any additional value added there?  (I must be tired -- don't take this the wrong way. I very much respect your work with this -- just want to avoid the changes where s=we don't need changes.)

I have one comments below about references, and the others are hopefully more focused on the MEO/PEO stuff.

Luis-Varona commented:
@dschult Don't worry, I'm certainly not taking anything personally! I really appreciate the feedback, and on second thought I do concur that it might be more prudent to create a separate PR to fix up the References bits. 🙂

Regarding the "self loop" -> "self-loop" change, I just know that self-loop is more common in the literature (and is probably just more grammatically correct).

Regarding the period-in-docstring stuff, I just noticed that some docstrings did have periods whereas others didn't, so I once again took it upon myself to standardize things (perhaps we should also just fold that into the References PR for this file?).

I plan to sleep soon-ish, but I'll look over the rest of your review/comments tomorrow. Thanks again!

amcandio reviewed (commented):
Just double checking, there is no behavior change in this PR right?  What is the advantage of having 3 methods as opposed to 1?

Luis-Varona commented:
@dschult and @amcandio, I've made the requested changes (including "folding" the MCS-M logic of `_triangulate_fully` into `minimal_elimination order`) and squashed into one commit. (At least, I've made the edits @dschult wanted; I'm not sure if @amcandio is already clear on why we need all three of `complete_to_chordal_graph`, `minimal_elimination_order`, and `perfect_elimination_order`.) Just let me know if there's anything else you'd like me to do before merging. 🙂

**PS:** The failing documentation/tests from the CI run are, I believe from a different part of the codebase completely unrelated to this PR.

Luis-Varona commented:
**PPS:** Once again, @dschult, refer to the Chicago Manual of Style, which `networkx` officially recommends (see https://networkx.org/documentation/latest/developer/contribute.html#adding-references), for why I formatted the authors in my added References as `LastName1, FirstName1, FirstName2 LastName2, and FirstName3 LastName3`. (In the very `networkx` contributing guidelines this is recommended.)

I just did this for the new functions and `is_chordal`, however, leaving the docstrings unrelated to this PR alone to minimize distractions. 🙂

dschult commented on networkx/algorithms/chordal.py:
```diff
@@ -395,35 +401,78 @@ def complete_to_chordal_graph(G):
+
+@not_implemented_for("directed")
+@not_implemented_for("multigraph")
+@nx._dispatchable
```
This should match the dispatchable decorator call that is used for `complete_to_chordal_graph` because this function returns a networkx graph. That'll fix one of the CI tests. The other two involve circleCI and I think are circleCI network related.

Luis-Varona replied:
Got it; I'll quickly fix that later this evening, probably. Anything else you'd like me to change, @dschult?

Luis-Varona commented:
@dschult, the dispatchable decorator bug has been fixed. Anything else you or @amcandio would like me to change before merging? Thanks again for the really helpful comments and timely reviews! 🙂

amcandio reviewed (approved):
Looks good to me

Luis-Varona commented:
@rossbar, I believe (if I'm not misinterpreting things) that it was @dschult who originally proposed this terminology in #6873. @dschult, your thoughts? Is this something we need to fix before merging?

Luis-Varona commented:
> @rossbar, I believe (if I'm not misinterpreting things) that it was @dschult who originally proposed this terminology in #6873. @dschult, your thoughts? Is this something we need to fix before merging?

See also [my comment](https://github.com/networkx/networkx/issues/6873#issuecomment-2889386659) back in #6873, @rossbar.

dschult commented:
I/we seem to have been distracted by the style changes instead of the content changes. Can you remove all but the content changes?

Meanwhile the discussion can focus on what we want for our interface.  In #6873 I proposed the following:
> So, it seems that we have many functions that could all use this underlying code:
>
> - [Edit: `triangulate` might be a better name] make_perfect_elimination_graph : return the graph created in current code
> - minimal_elimination_order : return the order given by alpha from the current code
> - perfect_elimination_order : if not is_chordal raise, else return MEO given by alpha
> - ~~is_perfect_elimination_graph :~~ is_chordal (no new function needed... perhaps add a sentence to docstring.

I think the current state of this PR is:
- minimum_elimination_order : returns H, alpha
- perfect_elimination_order : return alpha  (and requires a chordal input G)
- is_chordal : True/False whether G is a chordal graph which is equivalent to G having a perfect elimination order.

If I understand correctly, alpha contains the info to get an ordering of the nodes. But @rossbar is correct that this info is not a natural way to express an ordering of the nodes.  It makes more sense to return a list when we are talking about it as a ordering.  Is there something else you would use `alpha` for? What do the users of this function do with `alpha`?

We've don't have a way to get a triangulation of the graph without getting the dict `alpha` too. 

And I think we need to be clear on the defn of "perfect elimination order".  Is it true that removing the nodes in that order always maintains the chordal property for each of the sequence of graphs obtained?  The wikipedia definition of PEO talks about a node along with its neighbors which are later in the order form a clique.  But that isn't very clear.  I think it is true that removing the nodes in the PEO order means each removed node forms a clique with its neighbors when it is removed. Now, does that mean removing that node leaves the graph chordal? I think so.

Would using the terminology "triangulated" instead of "chordal" help here? Can someone help explain what this is all used for and how we can think about the terminology?  (along with the question about what is `alpha` actually used for?)

Luis-Varona commented:
@dschult I will read the rest tomorrow, but I have indeed removed all of the style changes and kept only the content changes. 🙂

Luis-Varona commented:
> I think the current state of this PR is:
> 
> * minimum_elimination_order : returns H, alpha
> * perfect_elimination_order : return alpha  (and requires a chordal input G)
> * is_chordal : True/False whether G is a chordal graph which is equivalent to G having a perfect elimination order.

Sort of, @dschult? Currently, we have

* `is_chordal(G)`: returns `True` if `G` is a chordal graph else `False` (equivalent to checking whether `G` has a PEO)
* `complete_to_chordal_graph(G)`: returns `H` (`G` triangulated to a chordal graph) and `alpha` (an MEO if `G` is not chordal; all zeros if `G` is already chordal)
* `minimal_elimination_order(G)`: returns `H` and `alpha` (an MEO if `G` is not chordal; a PEO&mdash;which is also an MEO&mdash;if `G` is already chordal)
* `perfect_elimination_order(G)`: returns `H` (a copy of `G` to avoid shared mutability) and `alpha` (a PEO), assuming that the input graph `G` is already chordal (raising an error otherwise)

I think it is clear here that `complete_to_chordal_graph` exhibits some buggy behavior when the input graph is already chordal (indeed, the motivation behind this whole issue/PR)&mdash;thus, I propose either deprecating it or modifying it so that it returns only the triangulation, not a new node ordering. (That is, `complete_to_chordal_graph(G)` would return only `H`, not `alpha`.) Users who desire MEOs/PEOs can then turn to `minimal_elimination_order` and/or `perfect_elimination_order` instead. However, I don't know how API-breaking this change to the return type of `complete_to_chordal_graph` would be.

Your thoughts, @dschult and @rossbar?

dschult commented:
Can you comment on `alpha`, whether that is a reasonable way to indicate an ordering, and/or how that is used by the user of this suite of functions?

Luis-Varona commented:
> Can you comment on `alpha`, whether that is a reasonable way to indicate an ordering, and/or how that is used by the user of this suite of functions?

Will look into it. 🙂

dschult commented:
> whether that is a reasonable way to indicate an ordering
> 
> @dschult, may I ask precisely what you mean by this?

I am thinking that an ordering would naturally be provided as a list (or sequence of some type). My understanding is that `alpha` provides an ordering via a dict keyed by node to a value. Sorting the values gives the ordering. Is this understanding correct?

Do users find the ordering by sorting the keys of the dict `alpha` based on the values in the dict? Are there other uses for the values in `alpha`? 

It seems more natural to provide the peo/meo as a list. I'm trying to figure out if we lose anything by doing that. What is alpha actually used for by users?

Luis-Varona commented:
Thanks for the clarification, @dschult; it's been a busy few days at work, but I'll look into it over the weekend. I will say that I was not involved in the design decisions for `complete_to_chordal_graph` prior to this PR, so I am also somewhat unclear as to why `alpha` is given the way that it is and for what it is used. Nevertheless, I'll hopefully be able to figure something out sometime soon. 🙂

@rossbar, do you have any further thoughts on this matter?

rossbar commented:
I'm not familiar with the development history of `complete_to_chordal_graph` either. My guess is the fact that `alpha` is structured as a dict was dictated by the implementation rather than any UX design considerations.

Luis-Varona commented:
> I am thinking that an ordering would naturally be provided as a list (or sequence of some type). My understanding is that `alpha` provides an ordering via a dict keyed by node to a value. Sorting the values gives the ordering. Is this understanding correct?

Yes, that's exactly it. `alpha` is a dict because the nodes of our original `networkx` graph might not be labelled precisely `0, 1, …, n`. They might be other integers (or any object, really), so it makes more sense to treat `alpha` less as a pure "ordering" and more as a relabelling of these nodes.

> Do users find the ordering by sorting the keys of the dict `alpha` based on the values in the dict? Are there other uses for the values in `alpha`?

That's exactly right. As for other uses, I'm not an expert on the topic, but I believe the MEO of `G` gives us (indirectly in some cases, but more quickly than applying an algorithm from scratch):
- A tree decomposition of the triangulated (chordal) graph, `H`
- Data useful for some graph coloring heuristics applicable to `G`
- The treewidth of `H` and, by extension, a bound on the treewidth of `G`
- Probably a few more things

> It seems more natural to provide the peo/meo as a list. I'm trying to figure out if we lose anything by doing that. What is alpha actually used for by users?

I guess the main argument against that (besides the fact that we'd be potentially breaking API compatibility if we changed how `alpha` is structured) is that it's easier to go from the current dict from to list form than vice versa. If we provide `alpha` as a list and the nodes of our original `networkx` graph are not `0, 1, …, n`, then it becomes unnecessarily tricky to remap all the values back for the aforementioned tree decomposition of `H`, coloring heuristics for `G`, etc.

Anything else you'd like to investigate, @dschult? Do you think this is ready to merge or would you like to deliberate a bit further and change stuff?

Luis-Varona commented:
I do think as everything currently stands, the added MEO/PEO changes can't hurt at all. _Maybe_ the terminology, dict vs. list, etc. merits further consideration, but they would be additional improvements over the current ones. If we do change from dict vs. list, I'm also not sure whether that should be included in this PR or another, considering that change would actually break API compatibility. (And again, I stand by a dict being more natural as stated in my last comment given the potential uses of the MEO&mdash;although I wasn't even the one who implemented it this way in the first place.)

Do correct me if I'm wrong, though, @dschult, and let me know where you'd like to go from here. 🙂

Luis-Varona commented:
@dschult I just realized when you meant "Provide `alpha` as a list," you didn't mean "Provide the values (integers from 0 to `n - 1` sorted in the appropriate order," you meant "Provide (a subset of the) original nodes sorted in the appropriate order." My bad. Anyway, does some of the information I've provided help at all?

dschult commented:
Yes, this information helps. Thanks!

Regarding returning a list of nodes instead of a dict of nodes to alpha values:
- we lose the values of this alpha measure. But that doesn't seem to be a big deal. Those values aren't used for anyhing we know about right?
- we gain in that users don't have to role their own idiom to find the meo/peo from a dict of values.
- we gain simplicity of the return values. The meo/peo routines can just return a list of nodes. Only `complete_to_chordal_graph` would actually return a dict of nodes to alpha values (for backward compatibility) and that might get deprecated.

Any thoughts on moving to this kind of API?

Luis-Varona commented:
By the way, @dschult&mdash;do you concur that an immediate follow-up PR fixing the Chicago-style references once #8024 is merged would be appropriate? I can have that all ready to go as well if you acknowledge the need.

Luis-Varona commented:
PS: The most recent push doesn't make the desired change yet. I just wanted to update with the current main, reducing the risk of any merge conflicts. I'll get started on changing `alpha` to a list in `minimal_elimination_order` and `perfect_elimination_order` unless there are any objections, @dschult and @rossbar, leaving the current commit untouched in case we ever want to revert for some reason.

Luis-Varona commented:
By the way&mdash;I suppose as we near completion of this PR, we should also update `test_chordal.py` accordingly.

rossbar commented:
I'm still not sold on the proposed interfaces to this functionality. To me, it's not very natural to think of `complete_to_chordal_graph` in terms of elimination ordering. Setting aside backward compatibility constraints for the moment, to me the most natural interface is something like:

```python
def complete_to_chordal_graph(G):
"""
Returns
-------
set of edges
    A set of edges that upon addition to `G` transform
    `G` to a chordal graph. Returns empty if `G` is already
    chordal.
"""

def perfect_elimination_ordering(G):
"""
Returns
-------
list of nodes
    A perfect elimination ordering of `G`

Raises
------
NetworkXError
    If `G` is not chordal
"""
```

I understand that the MEO reflects the underlying implementation, but the `(modified graph, PEO of modified graph)` return idiom feels like it's providing information in such a way that a user has to do extra processing to get at the "core" information; i.e. the set of edges that triangulate `G` to a chordal graph.

rossbar commented:
> triangulation (via MCS-M) is needed to to obtain the MEO/PEO, so I suppose the point of returning both pieces of information is to avoid redundancy later on.

I think this is perhaps part of it - we're letting the implementation dictate (or at least influence) the interface. Admittedly this is entirely subjective, but I still find it more natural to think about perfect elimination orderings and chordal graphs than minimal elimination orderings of non-chordal graphs (where the triangulation-to-chordal happens under the hood).

In terms of redundancy, I think think that's subjective too. For example, with the current interface I can get an MEO straightaway, but if the question I want to ask is: "what's the minimum number of edges I need to add to `G` to make the resulting graph chordal", then I have to do extra processing:

```python
>>> H, meo = nx.complete_to_chordal_graph(G)
>>> chordalizing_edges = set(H.edges) - set(G.edges)
```

Both questions are equally valid - the "redundancy" in the just depends on the nature of the analysis. My (admittedly very much non-expert) opinion is that it's more natural to derive an MEO from a set of chordal-izing edges than vice versa.

> Anyway, I think we certainly shouldn't change the complete_to_chordal_graph interface for compatibility reasons.

Agreed, at least not without deprecation cycles etc. However one possibility is to add new functionality, e.g. a new function `triangulate_to_chordal` which returns edge sets. Then you have 3 primary functions for dealing chordality-related concepts: `is_chordal`, `triangulate_to_chordal`, and `perfect_elimination_ordering`. A minimal elimination ordering is then just a PEO on a chordal-ized graph:

```python
>>> H = G.copy()
>>> H = nx.add_edges_from(nx.triangulate_to_chordal(G))
>>> meo = nx.perfect_elimination_ordering(H)
```

rossbar commented:
From my perspective this can be decomposed into two separate issues: 1) fixing `complete_to_chordal_graph` (which this PR does) and 2) expanding the chordal API to add MEO and PEO functions.

Addressing point 1 is completely unobjectionable to me and pure bug fix that should go in. However, I would advocate for a solution that *doesn't* expand the API, as I think that discussion is a bit more involved. In adding new functions, I think we run the risk of perpetuating some of the confusions related to the existing functionality.

Luis-Varona commented:
@rossbar I just want to make sure there's no misunderstanding here—this PR doesn't change the behavior of `complete_to_chordal_graph` at all. Mine and @dschult's proposed "bugfix" to the all-zeros `alpha` issue takes the form of the new MEO function. We keep the current behavior of `complete_to_chordal_graph` for backwards compatibility.

rossbar commented:
> Mine and @dschult's proposed "bugfix" to the all-zeros alpha issue takes the form of the new MEO function.

Understood, and I'm pushing back on this - I view the "alpha-all-zeros-when-G-chordal" case of `complete_to_chordal_graph` a clear bug and not behavior that should be preserved.

I'm also challenging the addition of `MEO` as public-facing terminology. From a quick search of the internet, the terms `perfect elimination ordering` and `chordal graph` are very widely used and accepted (e.g. have wikipedia articles). A search of "minimal elimination ordering" on the other hand returns links to scholarly articles that have few citations. Using this internet search as a proxy for "community consensus" on terminology, my supposition is that NetworkX's will be more discoverable and understandable if we present the functionality in terms of these widely-accepted concepts. Hence the desire to treat minimal elimination ordering as the result of a perfect elimination ordering on a graph that has been triangulated to chordal.

rossbar commented:
> @rossbar is this rehash in line with what you had in mind?

Indeed, thanks for the nice clear summary @dschult ! My only thoughts on the API-design front is that we could workshop the name `triangulation_edges`.

I think this name makes sense if we only expose it in the `chordal` namespace, i.e. the calling pattern is `nx.chordal.triangulation_edges`. However, if it's going to make it all the way out to the top-level namespace then I think it's worth having some reference to `chordal` in the function name itself, e.g. `nx.triangulate_to_chordal`. I would say `complete_to_chordal` but that's already been used "/.

Either way, not a major sticking point, just something to consider!

> I was trying to propose that none of the functions return alpha. The order/ordering function should return a list of nodes. So is_chordal returns boolean, triangulation_edges returns a list of edges and perfect_elimination... returns a list of nodes.

Agreed on this front - representing an ordering as a dict is not at all intuitive IMO. I vote for no more alphas, and no more multiple returns, just:

```python
nx.is_chordal(G) -> bool
nx.chordal.triangulation_edges(G) -> set of edges
nx.perfect_elimination_ordering(G) -> list of nodes  # raises if G isn't chordal
```

Luis-Varona commented:
@rossbar We could, alternatively, simply call the new function `chordal_triangulation_edges`. "Chordal triangulation" is a widely used term when "triangulation" on its own is not clear enough—see, for instance, [Berry et al. 2006](https://doi.org/10.1016/j.jalgor.2004.07.001).

Luis-Varona commented:
Hi, @dschult and @rossbar—just letting you know I plan to resume work on this PR sometime in the next few days 🙂

Luis-Varona commented:
> I'm inclined to close this one and focus on the work in #8024 if for no other reason than we managed to hammer out the API through extensive discussions in that PR. WDYT @akshitasure12 ?

Hi @rossbar, @dschult, and @akshitasure12--just want to mention that I'm still 100% planning on finishing up this PR. It's just been a hectic few weeks. Hoping to resume discussion and finish up the remaining tasks for sure.

@akshitasure12, do take a look and let us know if you agree with the current plan for the API refactor going forward.
